crates/api_common/src/request.rs

```diff
@@ -63,71 +59,61 @@ pub async fn fetch_link_metadata(
   let opengraph_data = extract_opengraph_data(&html_bytes, url)
     .map_err(|e| info!("{e}"))
     .unwrap_or_default();
-  let thumbnail =
-    extract_thumbnail_from_opengraph_data(url, &opengraph_data, generate_thumbnail, context).await;
-
   Ok(LinkMetadata {
     opengraph_data,
     content_type: content_type.map(|c| c.to_string()),
-    thumbnail,
   })
 }
 
-#[tracing::instrument(skip_all)]
-pub async fn fetch_link_metadata_opt(
-  url: Option<&Url>,
-  generate_thumbnail: bool,
-  context: &LemmyContext,
-) -> LinkMetadata {
-  match &url {
-    Some(url) => fetch_link_metadata(url, generate_thumbnail, context)
-      .await
-      .unwrap_or_default(),
-    _ => Default::default(),
-  }
-}
 /// Generate post thumbnail in background task, because some sites can be very slow to respond.
 ///
 /// Takes a callback to generate a send activity task, so that post can be federated with metadata.
+///
+/// TODO: `federated_thumbnail` param can be removed once we federate full metadata and can
+///       write it to db directly, without calling this function.
+///       https://github.com/LemmyNet/lemmy/issues/4598
 pub fn generate_post_link_metadata(
   post: Post,
   custom_thumbnail: Option<Url>,
+  federated_thumbnail: Option<Url>,
   send_activity: impl FnOnce(Post) -> Option<SendActivityData> + Send + 'static,
   local_site: Option<LocalSite>,
   context: Data<LemmyContext>,
 ) {
   spawn_try_task(async move {
-    // Decide if the thumbnail should be generated
-    let allow_sensitive = local_site_opt_to_sensitive(&local_site);
-    let page_is_sensitive = post.nsfw;
-    let allow_generate_thumbnail = allow_sensitive || !page_is_sensitive;
-    let do_generate_thumbnail =
-      allow_generate_thumbnail && custom_thumbnail.is_none() && post.thumbnail_url.is_none();
-
-    // Generate local thumbnail only if no thumbnail was federated and 'sensitive' attributes allow it.
-    let metadata = fetch_link_metadata_opt(
-      post.url.as_ref().map(DbUrl::inner),
-      do_generate_thumbnail,
-      &context,
-    )
-    .await;
-
-    // If its an image post, it needs to overwrite the thumbnail, and take precedence
-    let image_url = if metadata
+    let metadata = match &post.url {
+      Some(url) => fetch_link_metadata(url, &context).await.unwrap_or_default(),
+      _ => Default::default(),
```

Comment:
metadata should probably be an option, rather than a default. Option forces you to deal with it not being fetched. You might need to add an `if let Some(` below, but its pry safer.

Also why `await.unwrap_or_default`, shouldn't it just be await.

Reply:
This is also [how it was handled before](https://github.com/LemmyNet/lemmy/pull/4615/files#diff-0cd13a4a101c05a54d9ab789fa06c9d48ab61c0e1f5c9b9a178440e9ba3a5ef5L83-L85). What do you suggest we do if metadata cant be fetched? I dont think theres anything useful we could do differently.

Reply:
Looking at it again, I spose its fine since every field on that metadata is optional anyway.

Reply:
This could probably be made cleaner using something from the futures_util crate that works similarly to Option::map

Reply:
Good idea but it turns out even more verbose and harder to read than the current code:
```rust
    let metadata = OptionFuture::from(
      post
        .url
        .as_ref()
        .map(|url| fetch_link_metadata(url, &context)),
    )
    .await
    .map(Result::ok)
    .flatten()
    .unwrap_or_default();
```